### PR TITLE
packaging: fix issues in pre/postremove scripts

### PR DESCRIPTION
**Pull request by elfchief**
Fixes several issues with the pre/postremove scripts for both rpm and deb packages. Specifically:

For postremove:

- the postremove script now functions correctly (i.e. restarts consul after a package upgrade) on rpm-based systems (where `$1` is numeric rather than `purge` or `upgrade`)
- `systemctl daemon-reload` is called on package removal (rather than only on upgrade)
- calls `systemctl try-restart` instead of `systemctl restart`, which will only (re)start consul if it was already running when the upgrade happened.

For preremove:
- if the package is being completely uninstalled (rather than upgraded), stop consul before removing the package

Tested on CentOS 7.9 and Ubuntu 20.04, but should be applicable to all rpm and deb based distributions.

### CODE REVIEW

**Comment by elfchief** [comment]
Not sure why the envoy integration tests are failing, didn't touch anything that could remotely affect any actual code tests, so will assume just a flakey test unless told otherwise. (I can't see a way to just re-run the test, either, as a random user)

File: .release/linux/preremove
Comment (eculver) [inline .release/linux/preremove:3]: do we care about the `upgrade` case here too?

**Comment by eculver** [comment]
This all _looks_ correct but it's a little over my head with regards to the systemd and nfpm bits. I would have to defer to others who know more about these tools to safely approve, maybe @kpenfound? Otherwise, this looks good to me! The part about calling `try-restart` makes a lot of sense.

File: .release/linux/preremove
Comment (alinsavix) [reply .release/linux/preremove:3]: In this case, no, because we don't want to stop/disable anything if we're doing an upgrade -- the postremove handles restarting things, so we just let the upgrade case fall through. Normal package upgrade behavior, barring specific reasons to do things a different way, is "leave things running, swap out for the new version, then conditionally restart".

**Review by kpenfound — APPROVED** [review]
Yeah looks great @elfchief , thanks for the contribution! Definitely better handling of upgrade and remove than before

**Comment by mdeggies** [comment]
Want to do the honors of merging, @eculver ? :)

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/572746.

**Pull request merged**
